### PR TITLE
feat(auth): adds local state management 

### PR DESCRIPTION
**Pull request by israx**
#### Description of changes

This PR adds a local state management solution that is only used during the signIn flow.

The approach is inspired by the popular library Redux.

**Action items:**
- Remove this implementation and replace it by state machines.

#### Checklist

- [X] PR description included
- [x] `yarn test` passes
- [ ] Tests are [changed or added](https://github.com/aws-amplify/amplify-js/blob/main/CONTRIBUTING.md#steps-towards-contributions)
- [ ] Relevant documentation is changed or added (and PR referenced)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: packages/auth/src/providers/cognito/utils/signInStore.ts [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+			return state;
+	}
+};
+
+const createStore: Store<SignInState, SignInAction> = reducer => {
+	const initialSignInState = {
+		username: undefined,
+		activeChallengeName: undefined,
+		activeSignInSession: undefined,
+	};
```
Comment (jimblanc) [inline packages/auth/src/providers/cognito/utils/signInStore.ts:59]: Could we move this up to module scope in a utility function for single source of truth? Then on the next line we could just do `let currentState = defaultState()`

E.g.
```
const defaultState = (): SignInState => ({
  username: undefined,
  activeChallengeName: undefined,
  activeSignInSession: undefined,
})
```

**Review by jimblanc — COMMENTED** [review]
Makes sense to me as a temporary solution until the state machine is ready 🤔 Left a few comments on the state manager itself

File: packages/auth/src/providers/cognito/utils/signInStore.ts [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+// TODO: replace all of this implementation with state machines
+type SignInState = {
+	username: string | undefined;
+	activeChallengeName: string | undefined;
+	activeSignInSession: string | undefined;
+};
+
+type SignInAction =
+	| { type: 'SET_INITIAL_STATE' }
+	| { type: 'SET_USERNAME'; value: string | undefined }
```
[the remark this answers is not in the document]
Comment (israx) [reply packages/auth/src/providers/cognito/utils/signInStore.ts:15]: the intention was to require a value. There are cases were the values should be defined as `undefined`

**Review by jimblanc — APPROVED** [review]

**Review by jimblanc — APPROVED** [review]

**Review by AllanZhengYP — APPROVED** [review]
it looks great to me, nice!

**Pull request merged**